crates/bcr-ebill-transport/src/nostr.rs

```diff
@@ -470,112 +560,194 @@ impl NostrConsumer {
+                                &contact_service,
+                                &offset_store,
+                            )
+                            .await
+                        {
+                            // Determine which local identity should receive this event
+                            match determine_recipient(&event, &client).await {
+                                Ok((recipient_node_id, signer)) => {
+                                    let (success, time) = process_event(
+                                        event.clone(),
+                                        signer,
+                                        recipient_node_id.clone(),
+                                        chain_key_store,
+                                        &event_handlers,
+                                    )
+                                    .await?;
+                                    // store the new event offset for the recipient identity
+                                    add_offset(
+                                        &offset_store,
+                                        event.id,
+                                        time,
+                                        success,
+                                        &recipient_node_id,
+                                    )
                                     .await;
+                                }
+                                Err(e) => {
+                                    debug!(
+                                        "Could not determine recipient for event {}: {e}",
+                                        event.id
+                                    );
+                                }
                             }
-                            Ok(false)
                         }
-                    })
-                    .await
-                    .expect("Nostr notification handler failed");
-            });
-        }
+                        Ok(false)
+                    }
+                })
+                .await
+                .unwrap_or_else(|e| {
+                    error!("Nostr notification handler failed: {e}");
+                });
+        });
 
         Ok(tasks)
     }
 }
 
+/// Determines which local identity should receive this event.
+/// For private messages: tries to decrypt with each signer, returns the one that succeeds.
+/// For public chain events: the recipient is determined by chain key ownership (all identities have access).
+/// Returns the NodeId of the recipient identity and its signer.
+pub async fn determine_recipient(
+    event: &Event,
+    client: &NostrClient,
+) -> Result<(NodeId, Arc<dyn NostrSigner>)> {
+    match event.kind {
+        Kind::EncryptedDirectMessage | Kind::GiftWrap => {
+            // Try to decrypt with each identity's signer
+            // Clone the keys to avoid holding the lock during async operations
+            let keys_to_try: Vec<(NodeId, Arc<Keys>)> = client
```

Comment:
Hmm 😅 this likely won't scale well to a huge nomber of identities, right? But I guess n unwraps is better than the alternative of having n open connections all the time.